File: src/wilkins/dataflow.cpp
#include <wilkins/dataflow.hpp>

string
wilkins::
Dataflow::name()
{
    return name_;
}

string
wilkins::
Dataflow::execGroup()
{
    return execGroup_;
}

string
wilkins::
Dataflow::fullName()
{
    return fullName_;
}

int
wilkins::
Dataflow::in_passthru()
{
    return in_passthru_;
}

int
wilkins::
Dataflow::in_metadata()
{
    return in_metadata_;
}

int
wilkins::
Dataflow::out_passthru()
{
    return out_passthru_;
}

int
wilkins::
Dataflow::out_metadata()
{
    return out_metadata_;
}

int
wilkins::
Dataflow::zerocopy()
{
    return zerocopy_;
}

WilkinsSizes*
wilkins::
Dataflow::sizes()
{
    return &sizes_;
}

// whether this rank is producer, or consumer
bool
wilkins::
Dataflow::is_prod()
{
    return((type_ & WILKINS_PRODUCER_COMM) == WILKINS_PRODUCER_COMM);
}

bool
wilkins::
Dataflow::is_con()
{
    return((type_ & WILKINS_CONSUMER_COMM) == WILKINS_CONSUMER_COMM);
}

bool
wilkins::
Dataflow::is_prod_root()
{
    return world_rank_ == sizes_.prod_start;
}

bool
wilkins::
Dataflow::is_con_root()
{
    return world_rank_ == sizes_.con_start;
}

CommHandle
wilkins::
Dataflow::prod_comm_handle()
{
    return prod_comm_->handle();
}

CommHandle
wilkins::
Dataflow::con_comm_handle()
{
    return con_comm_->handle();
}

wilkins::
Dataflow::Dataflow(CommHandle world_comm,
                   int workflow_size,
                   int workflow_rank,
                   int& io_proc,
                   WilkinsSizes& wilkins_sizes,
                   int prod,
                   int dflow,
                   int con,
                   WorkflowLink wflowLink):
    world_comm_(world_comm),
    world_size_(workflow_size),
    world_rank_(workflow_rank),
    sizes_(wilkins_sizes),
    wflow_prod_id_(prod),
    wflow_dflow_id_(dflow),
    wflow_con_id_(con),
    type_(WILKINS_OTHER_COMM),
    tokens_(0),
    in_passthru_(0),
    in_metadata_(1),
    out_passthru_(0),
    out_metadata_(1),
    zerocopy_(0)
{

    // ensure sizes and starts fit in the world
    if (sizes_.prod_start + sizes_.prod_size > world_size_   ||
            sizes_.con_start + sizes_.con_size > world_size_)
    {
        fprintf(stderr, "Wilkins error: Group sizes of producer, consumer, and dataflow exceed total "
                        "size of world communicator\n");
        return;
    }

    //orc@13-07: lowfive related flags
    in_passthru_ = wflowLink.in_passthru;
    in_metadata_ = wflowLink.in_metadata;
    out_passthru_ = wflowLink.out_passthru;
    out_metadata_ = wflowLink.out_metadata;
    zerocopy_ = wflowLink.zerocopy;

    name_ = wflowLink.name;
    fullName_ = wflowLink.fullName;
    execGroup_ = wflowLink.execGroup;

    // communicator creation -- only applies to MPMD mode for the user codes
    if (!wilkins_master())
    {
        if (world_rank_ >= sizes_.prod_start &&                   // producer
                world_rank_ < sizes_.prod_start + sizes_.prod_size)
        {
            type_ |= WILKINS_PRODUCER_COMM;
            //orc@24-01: supporting subset of writers for prod
            if(sizes_.prod_writers==-1)
                prod_comm_ = new Comm(world_comm, sizes_.prod_start, sizes_.prod_start + sizes_.prod_size - 1);
            else if(world_rank_ < sizes_.prod_start + sizes_.prod_writers)
                prod_comm_ =  new Comm(world_comm, sizes_.prod_start, sizes_.prod_start + sizes_.prod_writers - 1);
            else
                io_proc = 0; //used by wilkins.py to determine which procs should join L5 ops

        }

        if (world_rank_ >= sizes_.con_start &&                    // consumer
                world_rank_ < sizes_.con_start + sizes_.con_size  && sizes_.con_start!=sizes_.prod_start) //orc@08-11: last if condition is to prevent duplicate comms in TP mode
        {
            type_ |= WILKINS_CONSUMER_COMM;
            con_comm_ = new Comm(world_comm, sizes_.con_start, sizes_.con_start + sizes_.con_size - 1);

            tokens_ = wflowLink.tokens;
        }

    }

}

wilkins::
Dataflow::~Dataflow()
{

    if (!wilkins_master())
    {
        if (is_prod())
            delete prod_comm_;
        if (is_con())
            delete con_comm_;
    }

}

File: bindings/python/wilkins/dataflow.py
# ---------------------------------------------------------------------------
#
# dataflow interface -- Python port of include/wilkins/dataflow.hpp
#                       and src/wilkins/dataflow.cpp
#
# --------------------------------------------------------------------------

from __future__ import annotations

import sys

from mpi4py import MPI

from .types import (
    WILKINS_OTHER_COMM,
    WILKINS_PRODUCER_COMM,
    WILKINS_CONSUMER_COMM,
    WilkinsSizes,
)
from .comm import Comm
from .workflow import WorkflowLink
from .context import wilkins_master


class Dataflow:
    """Represents a single producer-consumer dataflow link.

    World is partitioned into {producer, dataflow, consumer, other} in
    increasing rank.
    """

    def __init__(
        self,
        world_comm,          # MPI communicator (mpi4py Comm)
        workflow_size: int,  # size of the workflow
        workflow_rank: int,  # rank in the workflow
        io_proc: list,       # mutable list [io_proc_value] -- pass-by-ref emulation
        wilkins_sizes: WilkinsSizes,
        prod: int,           # id in workflow structure of producer node
        dflow: int,          # id in workflow structure of dataflow link
        con: int,            # id in workflow structure of consumer node
        wflow_link: WorkflowLink,
    ):
        self._world_comm = world_comm
        self._world_size = workflow_size
        self._world_rank = workflow_rank
        self._sizes = WilkinsSizes(
            prod_size=wilkins_sizes.prod_size,
            prod_writers=wilkins_sizes.prod_writers,
            con_size=wilkins_sizes.con_size,
            prod_start=wilkins_sizes.prod_start,
            con_start=wilkins_sizes.con_start,
        )
        self._wflow_prod_id = prod
        self._wflow_con_id = con
        self._wflow_dflow_id = dflow
        self._type = WILKINS_OTHER_COMM
        self._tokens = 0

        # LowFive related flags
        self._in_passthru = wflow_link.in_passthru
        self._in_metadata = wflow_link.in_metadata
        self._out_passthru = wflow_link.out_passthru
        self._out_metadata = wflow_link.out_metadata
        self._zerocopy = wflow_link.zerocopy
        self._flow_policy = wflow_link.flow_policy

        self._name = wflow_link.name
        self._full_name = wflow_link.fullName
        self._exec_group = wflow_link.execGroup

        self._prod_comm = None
        self._con_comm = None

        # Ensure sizes and starts fit in the world
        if (self._sizes.prod_start + self._sizes.prod_size > self._world_size or
                self._sizes.con_start + self._sizes.con_size > self._world_size):
            print(
                "Wilkins error: Group sizes of producer, consumer, and dataflow "
                "exceed total size of world communicator",
                file=sys.stderr,
            )
            return

        # Communicator creation -- only applies to MPMD mode for the user codes
        if not wilkins_master():
            # Producer
            if (self._world_rank >= self._sizes.prod_start and
                    self._world_rank < self._sizes.prod_start + self._sizes.prod_size):
                self._type |= WILKINS_PRODUCER_COMM

                # Supporting subset of writers for prod
                if self._sizes.prod_writers == -1:
                    self._prod_comm = Comm(
                        world_comm,
                        self._sizes.prod_start,
                        self._sizes.prod_start + self._sizes.prod_size - 1,
                    )
                elif self._world_rank < self._sizes.prod_start + self._sizes.prod_writers:
                    self._prod_comm = Comm(
                        world_comm,
                        self._sizes.prod_start,
                        self._sizes.prod_start + self._sizes.prod_writers - 1,
                    )
                else:
                    # Used by wilkins.py to determine which procs should join L5 ops
                    io_proc[0] = 0

            # Consumer
            # Last condition prevents duplicate comms in TP mode
            if (self._world_rank >= self._sizes.con_start and
                    self._world_rank < self._sizes.con_start + self._sizes.con_size and
                    self._sizes.con_start != self._sizes.prod_start):
                self._type |= WILKINS_CONSUMER_COMM
                self._con_comm = Comm(
                    world_comm,
                    self._sizes.con_start,
                    self._sizes.con_start + self._sizes.con_size - 1,
                )
                self._tokens = wflow_link.tokens

    def __del__(self):
        # In MPMD mode, free created communicators
        if not wilkins_master():
            # Comm objects handle their own cleanup via __del__
            pass

    def sizes(self) -> WilkinsSizes:
        """Return the sizes struct."""
        return self._sizes

    def is_prod(self) -> bool:
        """Whether this rank is a producer."""
        return (self._type & WILKINS_PRODUCER_COMM) == WILKINS_PRODUCER_COMM

    def is_con(self) -> bool:
        """Whether this rank is a consumer."""
        return (self._type & WILKINS_CONSUMER_COMM) == WILKINS_CONSUMER_COMM

    def is_prod_root(self) -> bool:
        """Whether this rank is the producer root."""
        return self._world_rank == self._sizes.prod_start

    def is_con_root(self) -> bool:
        """Whether this rank is the consumer root."""
        return self._world_rank == self._sizes.con_start

    def prod_comm_handle(self):
        """Return the producer MPI communicator."""
        return self._prod_comm.handle()

    def con_comm_handle(self):
        """Return the consumer MPI communicator."""
        return self._con_comm.handle()

    def in_passthru(self) -> int:
        """Consumer passthru flag."""
        return self._in_passthru

    def in_metadata(self) -> int:
        """Consumer metadata flag."""
        return self._in_metadata

    def out_passthru(self) -> int:
        """Producer passthru flag."""
        return self._out_passthru

    def out_metadata(self) -> int:
        """Producer metadata flag."""
        return self._out_metadata

    def flowPolicy(self) -> int:
        """Flow control policy."""
        return self._flow_policy

    def zerocopy(self) -> int:
        """Zerocopy flag."""
        return self._zerocopy

    def name(self) -> str:
        """Link name."""
        return self._name

    def execGroup(self) -> str:
        """Execution group name."""
        return self._exec_group

    def fullName(self) -> str:
        """Full link name with source."""
        return self._full_name
